core/codec/uvarint.hpp

```diff
@@ -3,25 +3,86 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef CPP_FILECOIN_CODEC_UVARINT_HPP
-#define CPP_FILECOIN_CODEC_UVARINT_HPP
+#pragma once
 
 #include <gsl/span>
-#include <libp2p/multi/uvarint.hpp>
+#include <limits>
 
+#include "common/file.hpp"
 #include "common/outcome.hpp"
 
 namespace fc::codec::uvarint {
   using Input = gsl::span<const uint8_t>;
 
+  struct VarintDecoder {
+    size_t max_bits{64};
+    uint64_t value{0};
+    bool more{true};
+    bool overflow{false};
+    size_t length{0};
+
+    inline void update(uint8_t byte) {
+      assert(more);
+      assert(!overflow);
+      more = byte & 0x80;
+      auto bits7{byte & 0x7f};
+      auto shift{length * 7};
+      auto more_bits{max_bits - shift};
+      overflow = more_bits < 7 && (bits7 & (-1 << more_bits));
+      value |= bits7 << shift;
+      shift += 7;
+      ++length;
+    }
+  };
+
+  /** returns true on success */
+  inline bool read(std::istream &is, VarintDecoder &varint) {
```

Comment:
May be make function a `VarintDecoder` method?

Reply:
No, free functions are better